libs/contrib/Control/Algebra/Laws.idr

```diff
@@ -0,0 +1,332 @@
+module Control.Algebra.Laws
+
+import Prelude
+import Control.Algebra
+
+%default total
+
+-- Monoids
+
+||| Inverses are unique.
+public export
+uniqueInverse : VMonoid ty => (x, y, z : ty) ->
+  y <+> x = neutral {ty} -> x <+> z = neutral {ty} -> y = z
+uniqueInverse x y z p q =
+  rewrite sym $ monoidNeutralIsNeutralL y in
+    rewrite sym q in
+      rewrite semigroupOpIsAssociative y x z in
+  rewrite p in
+    rewrite monoidNeutralIsNeutralR z in
+      Refl
+
+-- Groups
+
+||| Only identity is self-squaring.
+public export
+selfSquareId : VGroup ty => (x : ty) ->
+  x <+> x = x -> x = neutral {ty}
+-- selfSquareId x p =
+--   rewrite sym $ monoidNeutralIsNeutralR x in
+--     rewrite sym $ groupInverseIsInverseR x in
+--   rewrite sym $ semigroupOpIsAssociative (inverse x) x x in
+--     rewrite p in
+--       Refl
+
+||| Inverse elements commute.
+public export
+inverseCommute : VGroup ty => (x, y : ty) ->
+  y <+> x = neutral {ty} -> x <+> y = neutral {ty}
+-- inverseCommute x y p = selfSquareId (x <+> y) prop where
+--   prop : (x <+> y) <+> (x <+> y) = x <+> y
+--   prop =
+--     rewrite sym $ semigroupOpIsAssociative x y (x <+> y) in
+--       rewrite semigroupOpIsAssociative y x y in
+--     rewrite p in
+--       rewrite monoidNeutralIsNeutralR y in
+--         Refl
+
+||| Every element has a right inverse.
+public export
+groupInverseIsInverseL : VGroup ty => (x : ty) ->
+  x <+> inverse x = neutral {ty}
+-- groupInverseIsInverseL x =
+--   inverseCommute x (inverse x) (groupInverseIsInverseR x)
+
+||| -(-x) = x
+public export
+inverseSquaredIsIdentity : VGroup ty => (x : ty) ->
+  inverse (inverse x) = x
+-- inverseSquaredIsIdentity x =
+--   let x' = inverse x in
```

Comment:
Idris2 isn't able to use `x' = inverse x` in its normalization. The following also reproduces the issue (the copiler complains about `prf`):
```idris
let x' = inverse x
    prf = the (x' = inverse x) Refl
 in ...
```

Reply:
Based on what @edwinb suggested, we could aid Idris2 in the equality between `x'` and `inverse x` using a local function like:

```idris
||| -(-x) = x
public export
inverseSquaredIsIdentity : VGroup ty => (x : ty) ->
  inverse (inverse x) = x
inverseSquaredIsIdentity x =
  let
    lemma : (x' : ty) -> (x' = inverse x) -> inverse (inverse x) = x
    lemma x' prf =
      uniqueInverse
        x'
        (inverse (inverse x))
        x
        (rewrite sym prf in groupInverseIsInverseR x')
        (rewrite prf in groupInverseIsInverseR x)
   in lemma (inverse x) Refl
```

Reply:
I suggest we avoid leaning on workarounds like this. As a user, I don't want to have to work around quirks in the type system. Ideally I would be able to express what I want to express without feeling like I have to walk on eggshells to satisfy the typechecker.

Especially in this particular case, the existing proof is clean and elegant, and this change obscures that.